extensions/ql-vscode/src/helpers.ts

```diff
@@ -50,10 +78,58 @@ export function withProgress<R>(
         const increment = 100 * (step - progressAchieved) / maxStep;
         progressAchieved = step;
         progress.report({ message, increment });
-      }, token);
+      }, token, ...args);
     });
 }
 
+/**
+ * A generic wrapper for commands. This wrapper adds error handling and progress monitoring
+ * for any command.
+ *
+ * There are two ways to invoke the command task: with or without a progress monitor
+ * If progressOptions are passed in, then the command task will run with a progress monitor
+ * Otherwise, no progress monitor will be used.
+ *
+ * If a task is run with a progress monitor, the first two arguments to the task are always
+ * the progress callback, and the cancellation token. And this is followed by any extra command arguments
+ * (eg- selection, multiselection, ...).
+ *
+ * If there is no progress monitor, then only extra command arguments are passed in.
+ *
+ * @param commandId The ID of the command to register.
+ * @param task The task to run. If passing taskOptions, then this task must be a `ProgressTask`.
+ * @param progressOptions Optional argument. If present, then the task is run with a progress monitor
+ *                    and cancellation token, otherwise it is run with no arguments.
+ */
+export function commandRunner<R>(
+  commandId: string,
+  task: ProgressTask<R> | NoProgressTask,
+  progressOptions?: ProgressOptions
```

Comment:
I wonder whether it would be safer to split this into two variants:
- accept `task: ProgressTask<R>` and `progressOptions`
- accept `task: NoProgressTask` and no `progressOptions`.

That way it's obvious whenever we attempt to write new code that uses this.

Reply:
So, `commandRunner` and `commandRunnerWithProgress`?  I guess this would allow us to ensure we send the proper number of arguments for each style.

Reply:
That sounds fine. I don't know whether TS overloading rules will let you have the same name with different types and arities.

Reply:
It's actually a good thing that I changed to different names. It uncovered one instance when I was calling a command handler with the wrong arguments.